<a href="https://colab.research.google.com/github/Sam-ga-al/Google-Collab/blob/main/Reto1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Se procede con la carga de las librerias de trabajo.

In [ ]:
import numpy as np
import pandas as pd

1. Se procede con la carga de la base de datos

In [ ]:
nxl='/content/1. BD2_CarCrash_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB_SF = XDB[XDB['County'] == 'SAN FRANCISCO']

XDB_SF=XDB_SF[['Weekday','ViolCat','ClearWeather','Month','CrashType','Highway','Daylight']]
XDB_SF['CrashType'] = pd.factorize(XDB_SF['CrashType'])[0]

XD=XDB_SF[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight']]
yd=XDB_SF[['Highway']]


2. Implementamos el modelo

In [ ]:
from re import U
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd)

ct=mnb.class_count_
print("El total de los datos es\n:",len(XD))
print("Porcentaje de los datos que hacen parte de una de las categorias de clasificacion es\n:",ct/len(XD))

#Para obtener la informacion del modelo
U=mnb.theta_
sigma=np.sqrt(mnb.var_)
cv=U/sigma
LI=U-sigma; LS=U+sigma
ndc=mnb.class_count_
pdnc=mnb.class_prior_  # % de los datos por categoria

print("Titulos")
print("La media de los datos es\n:",U)
print("Los limites superiores son\n:",U+sigma)
print("Los limites inferiores son\n:",U-sigma)



#Para evaluar una persona con las siguientes caracteristicas
XS=[[5,1,1,3,4,0]]
mnb.predict(XS)



El total de los datos es
: 2081
Porcentaje de los datos que hacen parte de una de las categorias de clasificacion es
: [0.75925036 0.24074964]
Titulos
La media de los datos es
: [[3.88417722 7.45696203 0.86265823 4.35189873 3.21772152 0.66265823]
 [3.9261477  4.1497006  0.79840319 6.69461078 2.27744511 0.62075848]]
Los limites superiores son
: [[ 5.83480263 10.978332    1.20686605  6.69526323  5.09144114  1.13546084]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  3.33509959  1.10595679]]
Los limites inferiores son
: [[1.9335518  3.93559205 0.51845041 2.00853424 1.3440019  0.18985561]
 [1.91011047 1.98771158 0.39721054 3.25297437 1.21979063 0.13556017]]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


array([0])

3. Se procede con la evaluacion de desemepño del modelo

In [ ]:
ydp=mnb.predict(XD)
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

#Exactitud
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
Ex = (VN+VP)/(VN+VP+FP+FN)
print("La exactitud es:",Ex)

#Tasa de error
TE=(FP+FN)/(VN+VP+FP+FN)
print("La tasa de error es:",TE)

#Sensibilidad
SEN=VP/(VP+FN)
print("La sensibilidad es:",SEN)

#Especificidad
Esp=VN/(VN+FP)
print("La espcificidad es:",Esp)

#Presicion
Pre=VP/(VP+FP)
print("La presicion es:",Pre)


[[1365  215]
 [ 224  277]]
La exactitud es: 0.7890437289764536
La tasa de error es: 0.21095627102354636
La sensibilidad es: 0.5528942115768463
La espcificidad es: 0.8639240506329114
La presicion es: 0.5630081300813008


Analisis de resultados:
El modelo de Naive Bayes Gaussiano, implementado para predecir accidentes en carreteras usando las variables dadas, muestra un rendimiento aceptable. Su exactitud es de un 78%, prediciendo correctamente la ocurrencia de accidentes. Sin embargo, la tasa de error (21%) indica que hay margen de mejora. La capacidad del modelo para identificar correctamente los accidentes reales (sensibilidad) es de (55%), mientras que su capacidad para identificar correctamente los casos sin accidentes (especificidad) es de 86%. La precisión, que refleja la proporción de predicciones positivas correctas, se sitúa en 56%.Por otro lado aunque la exactitud general es aceptable, la sensibilidad y la precisión podrían optimizarse   para un rendimiento más fiable en la identificación de accidentes.